### PR TITLE
MRG+1: update raw times

### PR DESCRIPTION
**Pull request by jasmainak**
closes https://github.com/mne-tools/mne-python/issues/3711

should we add an extra `param` to `raw.plot`? I think it's almost a bugfix ...

### CODE REVIEW

File: doc/whats_new.rst [hunk 1]
```diff
@@ -44,6 +44,8 @@ BUG
 
     - Fix channel name comparison in :func:`mne.read_montage` so that if ``ch_names`` is provided, the returned montage will have channel names in the same letter case by `Jaakko Leppakangas`_
 
+    - Fix :meth:`mne.io.Raw.plot` to use x-axis relative to `first_samp` by `Mainak Jas`_
```
Comment (Eric89GXL) [inline doc/whats_new.rst]: double-backticks on `first_samp`

File: mne/viz/raw.py [hunk 2]
```diff
@@ -822,8 +830,10 @@ def _plot_raw_traces(params, color, bad_color, event_lines=None,
             params['ax'].text((start + end) / 2., ylim[0], dscr, ha='center')
 
     # finalize plot
-    params['ax'].set_xlim(params['times'][0],
-                          params['times'][0] + params['duration'], False)
+    first_time = params['raw'].first_samp / float(sfreq)
```
Comment (Eric89GXL) [inline mne/viz/raw.py]: given then number of times `raw.first_samp / raw.info['sfreq']` shows up in this PR alone (at least 6), we might want to add a convenience private parameter as `raw._first_time`

**Comment by jasmainak** [comment]
okay, after discussing with @jaeilepp it seems that this is more complicated than I thought ...

One particularly pathological case is when you have concatenated raws. In that case, you have to reconstruct the `times` taking into account the different `raw._first samps`. You'd probably have to have a clear demarcation between the concatenated raws to make it work. Not sure if it's worth the effort.

**Comment by Eric89GXL** [comment]
I wouldn't worry about that case. I'd just worry about the `first_samp` of the first file. We have things like `concatenate_events` that should work fine with this approach (I think).

File: mne/viz/ica.py [hunk 3]
```diff
@@ -787,6 +787,8 @@ def _plot_sources_raw(ica, raw, picks, exclude, start, stop, show, title,
     inds = list(range(len(picks)))
     data = np.array(data)
     n_channels = min([20, len(picks)])
+    raw._first_time = raw.first_samp / float(raw.info['sfreq'])
```
Comment (Eric89GXL) [inline mne/viz/ica.py]: we shouldn't monkey-patch an object just for this. It's not very expensive to recompute it below

Comment (Eric89GXL) [inline mne/viz/ica.py]: ...if you really want this for convenience, just add a `@property` for `_BaseRaw._first_time`

Comment (jasmainak) [inline mne/viz/ica.py]: @Eric89GXL perhaps I misunderstood something in your comment [here](https://github.com/mne-tools/mne-python/pull/3729#discussion_r86401457). Isn't this what you were suggesting?

Comment (Eric89GXL) [inline mne/viz/ica.py]: I wasn't suggesting monkey-patching, no. A `@property` would be the cleanest way to do it:
```
@property
def _first_time(self):
    return self._first_samp / float(self.info['sfreq'])
```

Comment (jasmainak) [inline mne/viz/ica.py]: got it. I agree it's cleaner

**Comment by Eric89GXL** [comment]
Ready for merge from your end? `raw.plot()` and `ica.plot_sources(raw)` both seem to work for me and code LGTM. +1 for merge from me

**Comment by jasmainak** [comment]
yep, ready for merge!

**Comment by jaeilepp** [comment]
Annotation plotting is broken. And there's no way of plotting from 0 now?

**Comment by jasmainak** [comment]
I fixed annotation plotting (I think). Take a look @jaeilepp. Let me know if you want to have a parameter to allow plotting the old way.

**Comment by agramfort** [comment]
@jasmainak you need to rebase

@jaeilepp let us know if it is ok for you

thx

**Comment by jaeilepp** [comment]
Is this really something we want to do by default? It's not what ``mne_browse_raw`` does. At the very least there should be a way to disable it.

**Comment by Eric89GXL** [comment]
> It's not what mne_browse_raw does. At the very least there should be a way to disable it.

Good point, +1 for an option. I suppose it should default to the old behavior, too.

**Comment by jasmainak** [comment]
okay cool. I'll create an option !

**Comment by Eric89GXL** [comment]
@jasmainak any time to finish this up in the next week or so?

**Comment by jasmainak** [comment]
yo, I'll try to get to it today!

File: mne/viz/ica.py [hunk 4]
```diff
@@ -63,6 +64,8 @@ def plot_ica_sources(ica, inst, picks=None, exclude=None, start=None,
         Whether to halt program execution until the figure is closed.
         Useful for interactive selection of components in raw and epoch
         plotter. For evoked, this parameter has no effect. Defaults to False.
+    ignore_first_samp : bool
+        If False, show time axis relative to the ``raw.first_samp``.
```
Comment (Eric89GXL) [inline mne/viz/ica.py]: Double negatives are tough, how about `show_first_samp=False` by default, and `show_first_samp=True` means show times relative to `first_samp`?

**Comment by jasmainak** [comment]
okay @Eric89GXL comments addressed. Let's hope CIs are happy

**Comment by Eric89GXL** [comment]
LGTM, +1 for merge

Anyone else have time to look and merge?

File: mne/preprocessing/ica.py [hunk 5]
```diff
@@ -1380,10 +1380,12 @@ def plot_properties(self, inst, picks=None, axes=None, dB=True,
...
     @copy_function_doc_to_method_doc(plot_ica_sources)
     def plot_sources(self, inst, picks=None, exclude=None, start=None,
-                     stop=None, title=None, show=True, block=False):
+                     stop=None, title=None, show=True, block=False,
+                     show_first_samp=False):
         return plot_ica_sources(self, inst=inst, picks=picks, exclude=exclude,
                                 start=start, stop=stop, title=title, show=show,
-                                block=block)
+                                block=block,
+                                show_first_samp=show_first_samp)
```
Comment (jaeilepp) [inline mne/preprocessing/ica.py]: No need for a line break here.

**Comment by jasmainak** [comment]
@jaeilepp how about now?

**Comment by jaeilepp** [comment]
For some reason the time stamp for vertical line is not visible when ``show_first_samp=True``. It should appear above the `help` button.
screenshot_2017-01-12_08-27-36

**Comment by jasmainak** [comment]
Fixed! Let's hope @jaeilepp CI is happy ;)

**Comment by Eric89GXL** [comment]
Looks like a good fix to me, still +1 to merge so @jaeilepp feel free to merge if you're happy

**Comment by jaeilepp** [comment]
Thx @jasmainak

**Pull request merged**